File: templates/Enterprise-Template/src/typescript/enterprise-bot/src/dialogs/shared/enterpriseDialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import { RecognizerResult } from 'botbuilder';
import { LuisRecognizer } from 'botbuilder-ai';
import { DialogContext } from 'botbuilder-dialogs';
import { BotServices } from '../../botServices';
import { TelemetryLuisRecognizer } from '../../middleware/telemetry/telemetryLuisRecognizer';
import { CancelDialog } from '../cancel/cancelDialog';
import { CancelResponses } from '../cancel/cancelResponses';
import { MainResponses } from '../main/mainResponses';
import { InterruptableDialog } from './interruptableDialog';
import { InterruptionStatus } from './interruptionStatus';

export class EnterpriseDialog extends InterruptableDialog {

    // Fields
    private readonly services: BotServices;
    private readonly cancelResponder: CancelResponses = new CancelResponses();

    constructor(botServices: BotServices, dialogId: string) {
        super(dialogId);

        this.services = botServices;
        this.addDialog(new CancelDialog());
    }

    protected async onDialogInterruption(dc: DialogContext): Promise<InterruptionStatus> {

        // Check dispatch intent.
        const luisService: TelemetryLuisRecognizer | undefined = this.services.luisServices.get(process.env.LUIS_GENERAL || '');
        if (!luisService) {
            return Promise.reject(
                new Error('The specified LUIS Model could not be found in your Bot Services configuration.')
                );
            } else {
            const luisResult: RecognizerResult = await luisService.recognize(dc.context);
            const intent: string = LuisRecognizer.topIntent(luisResult, undefined, 0.1);

            switch (intent) {
                case 'Cancel':
                return this.onCancel(dc);
                case 'Help':
                return this.onHelp(dc);
                default:
            }
        }

        return InterruptionStatus.NoAction;
    }

    protected async onCancel(dc: DialogContext): Promise<InterruptionStatus> {
        if (dc.activeDialog && dc.activeDialog.id !== CancelDialog.name) {
            // Don't start restart cancel dialog.
            await dc.beginDialog(CancelDialog.name);

            // Signal that the dialog is waiting on user response.
            return InterruptionStatus.Waiting;
        }

        // Else, continue
        return InterruptionStatus.NoAction;
    }

    protected async onHelp(dc: DialogContext): Promise<InterruptionStatus> {
        const view: MainResponses = new MainResponses();
        view.replyWith(dc.context, MainResponses.responseIds.Help);

        // Signal the conversation was interrupted and should immediately continue.
        return InterruptionStatus.Interrupted;
    }
}

File: templates/Enterprise-Template/src/csharp/EnterpriseBotSample/EnterpriseBotSample/Dialogs/Shared/EnterpriseDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Threading;
using System.Threading.Tasks;
using EnterpriseBotSample.Dialogs.Cancel;
using EnterpriseBotSample.Dialogs.Main;
using Luis;
using Microsoft.Bot.Builder.Dialogs;

namespace EnterpriseBotSample.Dialogs.Shared
{
    public class EnterpriseDialog : InterruptableDialog
    {
        protected const string LuisResultKey = "LuisResult";

        // Fields
        private readonly BotServices _services;
        private readonly CancelResponses _responder = new CancelResponses();

        public EnterpriseDialog(BotServices botServices, string dialogId)
            : base(dialogId)
        {
            _services = botServices;

            AddDialog(new CancelDialog());
        }

        protected override async Task<InterruptionStatus> OnDialogInterruptionAsync(DialogContext dc, CancellationToken cancellationToken)
        {
            // check luis intent
            _services.LuisServices.TryGetValue("general", out var luisService);

            if (luisService == null)
            {
                throw new Exception("The specified LUIS Model could not be found in your Bot Services configuration.");
            }
            else
            {
                var luisResult = await luisService.RecognizeAsync<General>(dc.Context, cancellationToken);
                var intent = luisResult.TopIntent().intent;

                // Only triggers interruption if confidence level is high
                if (luisResult.TopIntent().score > 0.5)
                {
                    // Add the luis result (intent and entities) for further processing in the derived dialog
                    dc.Context.TurnState.Add(LuisResultKey, luisResult);

                    switch (intent)
                    {
                        case General.Intent.Cancel:
                            {
                                return await OnCancel(dc);
                            }

                        case General.Intent.Help:
                            {
                                return await OnHelp(dc);
                            }
                    }
                }
            }

            return InterruptionStatus.NoAction;
        }

        protected virtual async Task<InterruptionStatus> OnCancel(DialogContext dc)
        {
            if (dc.ActiveDialog.Id != nameof(CancelDialog))
            {
                // Don't start restart cancel dialog
                await dc.BeginDialogAsync(nameof(CancelDialog));

                // Signal that the dialog is waiting on user response
                return InterruptionStatus.Waiting;
            }

            // Else, continue
            return InterruptionStatus.NoAction;
        }

        protected virtual async Task<InterruptionStatus> OnHelp(DialogContext dc)
        {
            var view = new MainResponses();
            await view.ReplyWith(dc.Context, MainResponses.ResponseIds.Help);

            // Signal the conversation was interrupted and should immediately continue
            return InterruptionStatus.Interrupted;
        }
    }
}
